File: leetcode/904-fruits-into-baskets/main.go
package main

import "fmt"

func totalFruit(tree []int) int {
	seen := [40001]int{}
	var count, current, left int
	for right := 0; right < len(tree); right++ {
		if seen[tree[right]] == 0 {
			current++
		}
		seen[tree[right]]++
		for current > 2 {
			seen[tree[left]]--
			if seen[tree[left]] == 0 {
				current--
			}
			left++
		}
		if count < right-left+1 {
			count = right - left + 1
		}
	}
	return count
}

func main() {
	tree := []int{3, 3, 3, 1, 2, 1, 1, 2, 3, 3, 4}
	fmt.Println(totalFruit(tree))
}

File: leetcode/904-fruits-into-baskets/main.py
from typing import List
import collections


class Solution:
    def totalFruit(self, tree: List[int]) -> int:
        left, count, seen = 0, 0, collections.Counter()
        for right in range(len(tree)):
            seen[tree[right]] += 1
            while len(seen) > 2:
                seen[tree[left]] -= 1
                if not seen[tree[left]]:
                    seen.pop(tree[left])
                left += 1
                break
            count = max(count, right-left+1)
        return count


solution = Solution()
print(solution.totalFruit([3, 3, 3, 1, 2, 1, 1, 2, 3, 3, 4]))
